penguin/engine.py

```
@@ -2673,17 +2814,21 @@ async def _execute_codeact_actions(
         logger.debug(
             "[AUTO-CONTINUE FIX] Parsed %s actions from response", len(tool_calls)
         )
+        scheduled_tool_calls = tool_calls[:1]
+        for tool_call in scheduled_tool_calls:
+            self._persist_tool_call_record(cm, tool_call)
 
         async def _execute_actionxml_call(tool_call: Any) -> Any:
```

Ruff (0.15.12)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `tool_call`

(ANN401)